### ISSUE TITLE
Feature Request: Event to detect month change on datepicker popup

### ISSUE THREAD

**Issue by togakangaroo** [angular-ui/bootstrap#5409]
I need to do a query to the server to see which dates are available. There can be a lot so I don't want to query for all up-front. What I would like to do is tie into something like the bootstrap datepicker's `changeMonth` event to fire a query, but I don't see any such event exposed.

Ideally I should be able to do something along the lines of

```
                <input type="text" class="form-control"
                        placeholder="Date of appointment"
                        uib-datepicker-popup 
                        ng-model="dateOfAppointment"
                        is-open="isOpened"
                        change-month="onChangeMonth"
                        date-disabled="dateDisabled(date, mode)"
                        />
```

where in my directive I would have something like

```
let disabledTimeRanges = []
scope.dateDisabled = (date, mode) => 
          mode == 'day' && disabledTimeRanges.filter(x => x.start <= date && x.end <= date)[0]
scope.onChangeMonth = (date) =>
          $http.get(`/disabled-date-ranges/?near=${date}`).then(({data}) => 
              disabledTimeRanges.push(...data)
          )
```

**Comment by icfantv** [comment]
@togakangaroo, please do not use the issues forum for support-related requests.  Rather, please follow the instructions [here](https://github.com/angular-ui/bootstrap#need-help) as support-related questions are best served by and for the community.  The issues forum is reserved for bugs.  Thanks.

**Comment by togakangaroo** [comment]
@icfantv sorry I phrased this weird, it is meant to be a feature request, not a support request. I've looked at the code and as far as I can tell I'm pretty sure there is no way to do this.

**Comment by icfantv** [comment]
Ah, ok.  I will reopen if you indicate that it is a feature request and be more explicit on the behavior you would like.  Thanks.

**Comment by togakangaroo** [comment]
@icfantv rewritten

**Comment by icfantv** [comment]
I just want to be sure I understand exactly what you're asking.  You want to execute something when the user selects a date?

**Comment by icfantv** [comment]
If so, why don't you simply watch the bound model for changes and update whatever you intended accordingly?

**Comment by togakangaroo** [comment]
@icfantv I want to execute something when the user changes the selected *month*. So say I want to select April 1st last years

* I open the date picker
* I change the year to 2016
* I change the month to April
* I select the date from the calendar

Currently there's no way to hook into the second and third events.

**Comment by icfantv** [comment]
I understand.  Your best bet (and fastest) at this point, might be to override the month template `month.html`.  The reason I say this is I have no idea how the other devs would feel by adding yet another API  attribute to the datepicker.  They feel it's already way over-bloated.  @wesleycho, @Foxandxss, what do you guys think?  I could see a how this might be useful (and not just on the month buttons), but we've discuss the API bloat on datepicker a few times.

**Comment by wesleycho** [comment]
Why not disable the datepicker until the data is back? One shouldn't have the datepicker active while the data isn't back from the server, so it doesn't make sense to have it active and change how it behaves while the user is potentially interacting with it.

**Comment by icfantv** [comment]
He still needs the trigger, no?

**Comment by wesleycho** [comment]
When the data comes in, could just do ng-if or ng-show or something like that on another variable from the controller.

**Comment by icfantv** [comment]
Sorry, I believe he is saying he needs the trigger to fire off the request to get the data.

**Comment by wesleycho** [comment]
This sounds like it's better to have the data upfront if that is the case. I'm not sure we want to support this use case, as it seems like such an extreme optimization, and probably a minor one compared to many other things that can be done typically.

**Comment by icfantv** [comment]
I think he wants the callback to fire every time the month changes - so that he always has the latest code.

**Comment by togakangaroo** [comment]
@wesleycho well consider the use case I pointed out above; someone can look years into the past or the future - you really want to query and load up all periods on the page load?. It's really not a very extreme optimization at all, very reasonable for things like appointment bookings, time entry (where certain time periods can be locked - my use case), deliveries (can't take delivery when no one is scheduled to work the front desk), etc. I'd argue that this is maybe an intermediate use case at most as can be evidenced by the fact that bootstrap-datepicker includes this event in its API.

That being said, I understand the concern about api-bloat but you can get around that either via sensible defaults, or simply providing a hook into the underlying model where this event could be intercepted from.

**Comment by icfantv** [comment]
@togakangaroo, just a note about your rebuttal to the API bloat concern - both of those options you suggest add another feature to the API, whether or not it's used.  API hooks require tests.  API hooks require support and maintenance.

People think they want 1000s of options and 100% configurability until they actually get it and then it's totally unmanageable.  Both from a user perspective and from a development perspective.  That's API bloat - adding every conceivable feature request requested by our users.  I understand that you may not care, and rightly so, about any of those other features, rather just the ones that concern you.  We need to balance the needs of our users with the maintainability and usability of this API and in the end, simple and extensible is going to win out over corner case needs (not suggesting that yours is a corner case).

Instead of having the demo/documentation site like we do, we'd wind up with something like Angular's and it becomes insanely hard to find what you're looking for.

Also, don't forget that we're all volunteers here.  None of us get paid to do this.

**Comment by togakangaroo** [comment]
@icfantv I understand, I'm not upset and I appreciate what yall do, I just want to make sure the use case is clearly stated and not rejected out-of-hand as I really don't believe it to be an edge case.

**Comment by wesleycho** [comment]
One can get around that by using ng-if though was my point in my first comment, and avoid rendering the datepicker until the data is retrieved - would that not solve that issue?

**Comment by wesleycho** [comment]
Closing due to lack of comment & there being an adequate work around - we avoid supporting $scope eventing as much as possible due to perf overhead & awkward apis, so there has to be an extraordinary use case.

**Comment by togakangaroo** [comment]
@wesleycho that work around doesn't address the scenario I outlined where that is simply too much data to query (which for a calendar that can go years forward or back is very reasonable). 

I'm not sure what else I need to comment on. Is that simply a scenario you do not intend to support? That's fine but I would then request mentioning it front and center in the docs so people don't end up wasting time trying to get this to work with similar use cases.

For my specific case I was able to restructure my domain model so a single query will work but that's only because a very specific workflow in my app that enabled this.

**Comment by melino** [comment]
+1, but also for Datepicker (not Popup)

In my case I'd like to fetch `customClass` events from the backend, but only the relevant one for the current month view, as they are too many for all time.
It would be good to be able to update the `customClass` events each time the user changes the month.

**Comment by unloder** [comment]
+1, this is not some random feature, this is a likely use-case. jQuery bootstrap datepicker has these events, why can't angular ui have them then? 
I'm not very good at understanding angular, but i think you could do something similar to what angular google maps did. When you pass options to the directive, if you have an option "control", they bind a control object to this parameter that has links to the map object and some additional parameters and methods. Im not sure if it is in the docs or not but it is there and it is darn useful.
I don't know what trade-offs does this approach have, so please don't hate on me.

**Comment by wesleycho** [comment]
Please open a new issue so we can figure out an appropriate API. Eventing is not the answer, but maybe there are other ways we could wire in that would accomplish the same purpose but in a better fashion. Note that for customClass purposes, support for that will be blocked on investigating for dynamic class names.

**Comment by andyexeter** [comment]
+1 for this. There should be a way to know what month the datepicker is currently displaying, otherwise it is in an undefined state when viewing a different month to the current date's month.

To add to the use cases for this: I have a legend below a datepicker. I want to hide items within the legend if there are no relevant items in the days of the current month. Without being able to know the current _displayed_ month in the datepicker this seems to be impossible.

**Comment by ghost** [comment]
Hey @wesleycho, why closing the issue if the request wasn't even acknowledged, I mean seriously. How difficult can it be to understand a request for events API in which to capture active month browsed.
